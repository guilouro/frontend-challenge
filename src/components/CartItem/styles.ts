import styled, { css } from 'styled-components'
import * as InputField from 'components/InputField/styles'

export const Wrapper = styled.div`
  display: flex;
  width: 100%;
`

export const Image = styled.div`
  ${({ theme }) => css`
    margin-right: ${theme.spacings.xsmall};
  `}
`

export const Info = styled.div`
  flex: 1;
`

export const Item = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.normal};
    color: ${theme.colors.textLight};
    margin: ${theme.spacings.xsmall} 0;
  `}
`

export const Title = styled.h3`
  ${({ theme }) => css`
    color: ${theme.colors.text};
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.normal};
  `}
`
export const Select = styled(InputField.Input).attrs({ as: 'select' })`
  ${({ theme }) => css`
    width: 5rem;
    padding: 0.3rem ${theme.spacings.xxsmall};
  `}
`

export const RemoveButton = styled.div`
  cursor: pointer;
`
