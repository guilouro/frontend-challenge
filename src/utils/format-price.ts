export default function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR'
  }).format(price)
}
