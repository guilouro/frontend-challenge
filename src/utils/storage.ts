import { CartObject } from 'hooks/use-cart'

const APP_KEY = 'BARKYN_CART'

export function getStorageItem() {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(APP_KEY)
  return JSON.parse(data || '[]')
}

export function setStorageItem(value: CartObject) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(APP_KEY, data)
}
