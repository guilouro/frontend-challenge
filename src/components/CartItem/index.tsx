import Image from 'next/image'
import { ProductItemProps } from 'components/ProductItem'
import * as S from './styles'
import PromotionalPrice from 'components/PromotionalPrice'

export type CartItemProps = {
  quantity?: number
} & Omit<ProductItemProps, 'review'>

const CartItem = ({
  title,
  image,
  price,
  promotionalPrice,
  packageInfo,
  quantity = 1
}: CartItemProps) => (
  <S.Wrapper>
    <S.Image>
      <Image src={image} alt={title} width={64} height={64} loading="lazy" />
    </S.Image>
    <S.Info>
      <S.Item>
        <S.Title>{title}</S.Title>
        <span>X</span>
      </S.Item>
      <S.Item>
        <span>Price</span>
        <PromotionalPrice
          promotionalPrice={promotionalPrice}
          price={price}
          dark
        />
      </S.Item>
      <S.Item>
        <span>Package</span>
        <span>{packageInfo}</span>
      </S.Item>
      <S.Item>
        <label htmlFor="quantity">Quantity</label>
        <span>
          <input
            id="quantity"
            type="number"
            name="quantity"
            defaultValue={quantity}
            maxLength={2}
          />
        </span>
      </S.Item>
    </S.Info>
  </S.Wrapper>
)

export default CartItem
