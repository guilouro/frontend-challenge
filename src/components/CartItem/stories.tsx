import { Story, Meta } from '@storybook/react'
import CartItem, { CartItemProps } from '.'

export default {
  title: 'CartItem',
  component: CartItem,
  args: {
    title: 'Barkyn Cannabis Oil para c√£es',
    image: '/img/products/oil.png',
    price: 2.29,
    promotionalPrice: 2.24,
    packageInfo: '170g',
    quantity: 2
  }
} as Meta

export const Default: Story<CartItemProps> = (args) => (
  <div style={{ maxWidth: 400 }}>
    <CartItem {...args} />
  </div>
)
