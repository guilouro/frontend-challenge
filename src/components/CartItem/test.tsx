import 'intl.mock'
import { render, screen } from 'utils/test-utils'

import CartItem from '.'

const props = {
  title: 'Barkyn Cannabis Oil para c√£es',
  image: '/img/products/oil.png',
  price: 2.29,
  promotionalPrice: 2.24,
  packageInfo: '170g'
}

describe('<CartItem />', () => {
  it('should render correctly', () => {
    const { container } = render(<CartItem {...props} />)

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()

    expect(screen.getByRole('img', { name: props.title })).toBeInTheDocument()

    expect(screen.getByRole('spinbutton', { name: /quantity/i })).toHaveValue(1)

    expect(screen.getByText(props.promotionalPrice)).toBeInTheDocument()
    expect(screen.getByText(props.price)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render quantity based on props', () => {
    render(<CartItem {...props} quantity={4} />)

    expect(screen.getByRole('spinbutton', { name: /quantity/i })).toHaveValue(4)
  })
})
