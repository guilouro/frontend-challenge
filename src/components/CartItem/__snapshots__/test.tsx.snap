// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> should render correctly 1`] = `
.c8 {
  border: 2px solid #e7e7e9;
  padding: 0.8rem 1.6rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.c1 {
  margin-right: 1.6rem;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 1.4rem;
  font-weight: 400;
  color: #6c6b76;
  margin: 1.6rem 0;
}

.c4 {
  color: #383743;
  font-size: 1.6rem;
  font-weight: 400;
}

.c9 {
  width: 5rem;
  padding: 0.3rem 0.8rem;
}

.c5 {
  cursor: pointer;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  font-weight: 600;
  color: #5f2eff;
  color: #6c6b76;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  margin-right: 1.6rem;
  font-weight: 400;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  font-weight: 600;
  color: #5f2eff;
  color: #383743;
}

<div
  class="c0"
  data-cy="cart-item"
>
  <div
    class="c1"
  >
    <div
      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
    >
      <div
        style="box-sizing: border-box; display: block; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
        />
      </div>
      <img
        alt="Barkyn Cannabis Oil para cães"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </div>
  </div>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <h3
        class="c4"
      >
        Barkyn Cannabis Oil para cães
      </h3>
      <div
        aria-label="Remove Barkyn Cannabis Oil para cães"
        class="c5"
        role="button"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAiIGhlaWdodD0iMjAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </div>
      </div>
    </div>
    <div
      class="c3"
    >
      <span>
        Price
      </span>
      <div>
        <div
          class="c6"
        >
          2.29
        </div>
        <div
          class="c7"
        >
          2.24
        </div>
      </div>
    </div>
    <div
      class="c3"
    >
      <span>
        Package
      </span>
      <span>
        170g
      </span>
    </div>
    <div
      class="c3"
    >
      <label
        for="quantity"
      >
        Quantity
      </label>
      <span>
        <select
          class="c8 c9"
          id="quantity"
          min="1"
          name="quantity"
          type="number"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
        </select>
      </span>
    </div>
  </div>
</div>
`
